dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ESContentletAPIImpl.java

```diff
@@ -963,7 +966,45 @@ private IHTMLPage loadPageByIdentifier ( String ident, boolean live, User user,
         }
         return results;
     }
-
+    @Override
+    public Object getFieldValue(Contentlet contentlet, com.dotcms.contenttype.model.field.Field theField){
+      if(theField instanceof ConstantField){
+            contentlet.getMap().put(theField.variable(), theField.values());
+        return theField.values();
+      }
+      if(theField instanceof HostFolderField){
+        if(FolderAPI.SYSTEM_FOLDER.equals(contentlet.getFolder())){
+             return contentlet.getHost();
+        }else{
+             return contentlet.getFolder();
+        }
+      }else if(theField instanceof CategoryField){
+        Category category;
+        try {
+          category = catAPI.find(theField.values(), APILocator.getUserAPI().getSystemUser(), false);
+          // Get all the Contentlets Categories
+          List<Category> selectedCategories = catAPI.getParents(contentlet, APILocator.getUserAPI().getSystemUser(), false);
+          Set<Category> categoryList = new HashSet<Category>();
+          List<Category> categoryTree = catAPI.getAllChildren(category, APILocator.getUserAPI().getSystemUser(), false);
+          if (selectedCategories.size() > 0 && categoryTree != null) {
+              for (int k = 0; k < categoryTree.size(); k++) {
```

Comment:
If we use a Set we can reduce the cyclic complexity, such as 

Set<Category> categoryTree = new TreeSet ( (cat1, cat3) -> categ.getInode().compareTo(cat.getInode()))
for (Category categ : selectedCategories) {
                    if (categoryTree.contains (categ) {
                         categoryList.add(categoryTree.get(categ));
                      }
                  }